### PR TITLE
(WIP) Delete empty directories when namespaces are deleted

### LINKED ISSUE #7032: Namespace/Project leftovers
For some time we have known namespaces leaves a directory structure under a resource objects it creates.  In openshift this is 1:1 with projects.  So as projects come and go there is a lot of leftover directories in etcd.  O(10^3)   

We'll run a quick experiment to determine the space consumed by leftovers. 

/cc @abhgupta @derekwaynecarr @jeremyeder

### PR DESCRIPTION
**Pull request by jawnsy**
This change resolves #7032

### CODE REVIEW

**Comment by liggitt** [comment]
before this would work, we'd need the new namespace controller from upstream that uses deletecollection, and to update the origin namespace controller to use deletecollection

**Comment by jawnsy** [comment]
Hey Jordan,

We did some research and found that 300k empty directories takes 254MB in memory and 215MB on disk (etcd).  Even though this solution isn't perfect, we'd like to get this change into Origin as soon as possible.  This means we'll diverge from upstream Kubernetes, but that should be okay since eventually Kubernetes will switch to etcd 3.x which doesn't have this problem.

Can you review this request? I tested by making the same changes in upstream Kubernetes and saw that the changes work, for the most part (just excluding the kinds like pods that aren't using the etcd registry.)

cc @derekwaynecarr @abhgupta 

Jonathan

**Comment by liggitt** [comment]
that doesn't seem like a big enough impact to justify carrying this

**Comment by derekwaynecarr** [comment]
unit tests?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go [hunk 1]
```diff
@@ -264,6 +264,20 @@ func (h *etcdHelper) Delete(ctx context.Context, key string, out runtime.Object)
 }
 
 // Implements storage.Interface.
+func (h *etcdHelper) DeleteDir(ctx context.Context, directory string) error {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: return an error, don't log

**Comment by derekwaynecarr** [comment]
read about how to name commits when making upstream changes:

https://github.com/openshift/origin/blob/273458a405fad1f498ede39459f04eabe8187787/tools/rebasehelpers/commitchecker/validate.go#L34

https://github.com/openshift/origin/blob/master/HACKING.md#cherry-picking-an-upstream-commit-into-origin

**Comment by derekwaynecarr** [comment]
@liggitt - so I think its fair to have @jawnsy determine what kinds in openshift would actually benefit the change from the code that is actually in the code base today.  As you note, this helps with upstream, but does not yet fix the dangling pod and service folders.  Nothing downstream uses delete collection, so maybe he can get a PR to start moving the ball forward on that in parallel since I do not anticipate we will get the dynamic client in time for 3.2 since it has not yet merged in 1.2 kube.

**Comment by jawnsy** [comment]
I was struggling to answer @derekwaynecarr's question for a few days, so @liggitt took mercy on me and wrote some quick Go code to determine which resources need to be updated.

Here's the code for posterity:

```go
	for resource, handler := range storage {
		_, isLister := handler.(rest.Lister)
		_, isDeleter := handler.(rest.Deleter)
		_, isGracefulDeleter := handler.(rest.GracefulDeleter)
		_, isCollectionDeleter := handler.(rest.CollectionDeleter)
		if isLister && (isDeleter || isGracefulDeleter) && !isCollectionDeleter {
			fmt.Println(resource)
		}
	}
	os.Exit(1)
```

This outputs:

```
roles
clusterRoleBindings
projects
imageStreamTags
oAuthAuthorizeTokens
roleBindings
oAuthAccessTokens
clusterRoles
```

This is the list of things that have to be listed and deleted one-by-one. However, roles, clusterRoleBindings, projects, imageStreamTags, roleBindings, and clusterRoles are all virtual (they're backed by other resources) so we don't have to worry about them.

Thus the remaining real resources that need to be updated are oAuthAuthorizeTokens and oAuthAccessTokens to use the new generic etcd storage mechanism.

@abhgupta how should we proceed given the above?

**Comment by liggitt** [comment]
Also note that both of those types are not namespaced, so would not factor into per-namespace empty directory counts

**Comment by abhgupta** [comment]
So, for all of the project scoped resources, the current PR will be able to delete their leftover directories in etcd. Is my understanding correct?

Also, since this PR is not being submitted/accepted upstream, what is the cost of carrying this patch in origin and do we have consensus on carrying this patch in origin?

**Comment by liggitt** [comment]
No, that just means that all the namespaced resources have a deletecollection API method. The controller that cleans up origin resources has not been updated to use deletecollection yet. The controller that cleans up kube resources uses deletecollection for most, but not all kube resources.

**Comment by abhgupta** [comment]
Alright, so it doesn't really make sense to have this patch at our end till we have the namespaced resources in origin being cleaned up via the deletecollection API call. 

Agreed?

**Comment by derekwaynecarr** [comment]
This should make kube 1.2 if I ever get a chance to finally finish the last code review updates:
https://github.com/kubernetes/kubernetes/pull/21400

It will move to use delete collection for all things that support it upstream - which is everything but services.  After the next rebase, we need to assess using this controller in origin for origin resources or updating the existing controller to call delete collection for origin resources.  I think it should support reuse, but we cannot use discovery in OpenShift to find its namespaced resources yet so it would be a hard coded list of group kinds that get passed into controller.

**Comment by liggitt** [comment]
> everything but services

and pods, for graceful deletion purposes, right?

**Comment by derekwaynecarr** [comment]
@liggitt - pods now use delete collection.

per the latest rebase, using the upstream namespace controller to delete downstream OpenShift resources is listed as a p2 https://github.com/openshift/origin/pull/7975

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go [hunk 2]
```diff
@@ -459,6 +457,11 @@ func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, lis
 			return nil, err
 		}
 	}
+
+	// Try to delete the directory, which will only succeed if the directory is empty.
+	// We can ignore errors since the cost of an empty directory is relatively small.
+	e.Storage.DeleteDir(ctx, e.KeyRootFunc(ctx));
```
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: can we not call this when list options has provided a field selector or label selector?

delete collection lets you scope the set of items to delete....

Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: We should only call this when the ctx has no namespace.

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: has a namespace?

**Comment by derekwaynecarr** [comment]
https://github.com/openshift/origin/blob/master/pkg/project/controller/controller.go#L100

^^^

I am less concerned about changing the existing controller to call delete collection on a per call basis.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go (hunk 2 above)
Comment (abhgupta) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: Yes, that is correct. We only want to delete namespaced resources. The comment was probably a typo.

Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: sorry, yes, it was a typo.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go (hunk 1 above)
Comment (jawnsy) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: @liggitt  Wouldn't this make the behaviour of DeleteDir different from the rest of the methods, e.g. Watch?

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: you're right... the other places are weird, too

if ctx is required, we should return an error
if it's not required, we shouldn't log an error

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go [hunk 3]
```diff
@@ -534,12 +534,22 @@ func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, lis
 		}()
 	}
 	wg.Wait()
+
+	if listOptions.LabelSelector == nil && listOptions.FieldSelector == nil && api.NamespaceValue(ctx) != "" {
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: move this to the success case (just before we `return listObj, nil` below)

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: these can be nil or the result of `labels.Everything()`, etc... change to
```
anyLabels := listOptions.LabelSelector == nil || listOptions.LabelSelector.Empty()
anyFields := listOptions.FieldSelector == nil || listOptions.FieldSelector.Empty()
if anyLabels && anyFields {
  ...
}
```

Comment (jawnsy) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: Updated the PR, thanks for the review!

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go [hunk 4]
```diff
@@ -534,10 +534,20 @@ func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, lis
...
+
 	select {
 	case err := <-errs:
 		return nil, err
 	default:
+		anyLabels := listOptions.LabelSelector == nil || listOptions.LabelSelector.Empty()
+		anyFields := listOptions.FieldSelector == nil || listOptions.FieldSelector.Empty()
+
+		if anyLabels && anyFields && api.NamespaceValue(ctx) != "" {
+			// Try to delete the directory, which will only succeed if the directory is empty.
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: This actually isn't right in the case of graceful deletion, which leaves records in place during the graceful deletion period.

Comment (jawnsy) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: @liggitt How many resources are subject to this graceful deletion? And in that case, don't we just have a few extra empty directories left laying around? I think as long as we do a reasonable job cleaning up as much as we can, it's better than not having this change.

Similarly, restricting this to only happen when neither label nor field selectors are specified will miss some empty etcd directories as well.

I'm not clear on what we more we need to do to proceed here - can you elaborate? cc @abhgupta

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: @derekwaynecarr does everything get deleted gracefully by default on namespace cleanup?

Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: It's up to the Kind to determine if it supports graceful deletion, which is only known when `Delete` calls `BeforeDelete` on the strategy associated with that Kind.

I think this code could be improved some more to no longer ignore the object returned here:
https://github.com/openshift/origin/pull/7344/files/2c74fc04f78d01bf6fc4fff7d2ff35f6f0139006#diff-51031c94aa64233e569e007fdbf96633R528

If a non-nil object is returned with no error, then at least one of the objects is still undergoing graceful termination, and we can avoid calling delete directory.

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: Is non-nil a real signal that the delete was graceful? I thought some resources return the object on delete, even in non-graceful cases

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/cacher.go [hunk 5]
```diff
@@ -244,6 +244,11 @@ func (c *Cacher) Delete(ctx context.Context, key string, out runtime.Object) err
 }
 
 // Implements storage.Interface.
+func (c *Cacher) DeleteDir(ctx context.Context, directory string) error {
```
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/cacher.go]: I think this should take the following:

`DeleteDir(ctx context.Context, key string, typeName string)`

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go [hunk 6]
```diff
@@ -313,6 +313,20 @@ func (h *etcdHelper) Delete(ctx context.Context, key string, out runtime.Object)
 }
 
 // Implements storage.Interface.
+func (h *etcdHelper) DeleteDir(ctx context.Context, directory string) error {
```
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: I think this should take the following:

`DeleteDir(ctx context.Context, key string, typeName string)`

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go [hunk 7]
```diff
@@ -313,6 +313,20 @@ func (h *etcdHelper) Delete(ctx context.Context, key string, out runtime.Object)
...
+	if ctx == nil {
+		return errors.New("Context is nil")
+	}
+
+	directory = h.prefixEtcdKey(directory)
+
+	startTime := time.Now()
+	_, err := h.etcdKeysAPI.Delete(ctx, directory, &etcd.DeleteOptions{Dir: true})
+	metrics.RecordEtcdRequestLatency("delete", directory, startTime)
```
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: I think this should be

`metrics.RecordEtcdRequestLatency("delete-collection", typeName, ...)`

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go (hunk 4 above)
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go]: Ahh!  You are right.  I mistakenly just looked at the if block for graceful and missed the outer-block.  Need to see who is configuring with `ReturnDeletedObject` as true, and I am out of time to do that now.  Shockingly, I am nervous about delete code paths.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go (hunk 7 above)
Comment (jawnsy) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: @derekwaynecarr Is this the only place that we need the typeName, or did you have other intended uses for it in mind?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/cacher.go (hunk 5 above)
Comment (jawnsy) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/cacher.go]: @derekwaynecarr `key` seems like a more generic parameter name than `directory` - I'm curious why you prefer `key` in this case?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go (hunk 7 above)
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go]: i wanted to be able to distinguish delete directory differently than delete of item.

**Comment by jawnsy** [comment]
We are going to withdraw this PR for now, as given the current plan and estimated growth rate, we do not anticipate the extra memory usage to be problematic. We can always reopen this if it becomes necessary later.

**Pull request closed without merging**